src/goto-symex/goto_state.h

```diff
@@ -20,6 +20,9 @@ Author: Romain Brenguier, romain.brenguier@diffblue.com
 
 #include "renaming_level.h"
 
+// 20210122 Forward declaration required
```

Comment:
We generally don't date comments, in favour of using git blame / praise.  Maybe we should but at the moment we don't

Reply:
Also, why is it required?  It doesn't look necessary if you are deleting the code that uses it...

Reply:
Updated the comment to remove the date stamp. Also added more detail to the comment about the declaration of the child class inside the parent class. The parent class explicitly uses the child class, this was resolved before by declaring the constructor for the child class inline. I don't know why the child class needs to be used in the parent, this just removes the duplicate constructor declaration and handles the required (forward) declaration.

Reply:
Thank you for these updates, but can we please actually dig a little deeper here? It seems the `apply_condition` function is what uses `goto_symex_statet`, but then either 1) `apply_condition` could safely be used with a `goto_statet` instead of a `goto_symex_statet`, or 2) `apply_condition` should become a member function of `goto_symex_statet` instead of being a member of `goto_statet`. Thanks!

Reply:
The `apply_condition` code in `goto_statet` has special cases related to `goto_symex_statet` that are non-trivial to split from the parent class. To do this properly is a larger refactor that impacts several locations in the code and there are already comments related to this (e.g. see `symex_goto.cpp` lines 35-46).
I'm not comfortable taking on this larger refactor at the moment, specially since this PR is just to remove some redundant code.